File: sync/src/main/scala/com/fan/mysql/parser/PreviousGtidsParser.java
package com.fan.mysql.parser;


import com.fan.mysql.binlog.BinlogEventParser;
import com.fan.mysql.dbsync.BinlogContext;
import com.fan.mysql.dbsync.LogBuffer;
import com.fan.mysql.event.BinlogEvent;
import com.fan.mysql.event.EventHeader;
import com.fan.mysql.event.impl.PreviousGtidsEvent;
import com.fan.mysql.position.GtidSetPosition;
import com.fan.mysql.util.ByteUtil;

import java.math.BigInteger;
import java.util.Collection;

@SuppressWarnings("unused")
public class PreviousGtidsParser implements BinlogEventParser {

    @Override
    public BinlogEvent parse(LogBuffer buffer, EventHeader header, BinlogContext context) {
        PreviousGtidsEvent event = new PreviousGtidsEvent(header);
        StringBuilder sb = new StringBuilder();
        long sidNumberCount = buffer.getLong64();
        for (int i = 0; i < sidNumberCount; i++) {
            if (sb.length() > 0) {
                sb.append(",");
            }
            byte[] sourcId = buffer.getData(16);
            String uuidStr = parseServerId(sourcId);
            sb.append(uuidStr);
            long internalCount = buffer.getLong64();
            for (int j = 0; j < internalCount; j++) {
                BigInteger from = buffer.getUlong64();
                BigInteger to = buffer.getUlong64().subtract(BigInteger.valueOf(1));
                sb.append(":").append(from).append("-").append(to);
            }
        }
        event.setGtidSet(sb.toString());
        return event;
    }

    public static byte[] getBody(String gtidSet) {
        GtidSetPosition pos = new GtidSetPosition(gtidSet);
        // calculate body size
        int bodySize = 8;
        Collection<GtidSetPosition.UUIDSet> uuids = pos.getUUIDSets();
        for (GtidSetPosition.UUIDSet uuidSet : uuids) {
            bodySize += 16 + 8;
            bodySize += 16 * uuidSet.getIntervals().size();
        }
        // fill body
        byte[] body = new byte[bodySize];
        int offset = 0;
        ByteUtil.int8store(body, offset, uuids.size());
        offset += 8;
        for (GtidSetPosition.UUIDSet uuidSet : uuids) {
            byte[] uuidByte = parseUuid(uuidSet.getUUID());
            System.arraycopy(uuidByte, 0, body, offset, 16);
            offset += 16;
            ByteUtil.int8store(body, offset, uuidSet.getIntervals().size());
            offset += 8;
            for (GtidSetPosition.Interval interval : uuidSet.getIntervals()) {
                ByteUtil.int8store(body, offset, interval.getStart());
                offset += 8;
                ByteUtil.int8store(body, offset, interval.getEnd() + 1);
                offset += 8;
            }
        }
        return body;
    }

    private static byte[] parseUuid(String uuid) {
        byte[] source_id = new byte[16];
        int array_index = 0;
        int str_index = 0;
        while (str_index < uuid.length()) {
            char high_char = uuid.charAt(str_index++);
            if (high_char == '-') {
                continue;
            }
            byte high = char2Byte(high_char);
            char low_char = uuid.charAt(str_index++);
            byte low = char2Byte(low_char);
            source_id[array_index++] = (byte) (high * 16 + low);
        }
        return source_id;
    }

    private String parseServerId(byte[] sourcId) {
        StringBuilder sb = new StringBuilder();
        if (sourcId == null || sourcId.length <= 0) {
            return null;
        }
        for (int i = 0; i < sourcId.length; i++) {
            int v = sourcId[i] & 0xff;
            String hv = Integer.toHexString(v);
            if (hv.length() < 2) {
                sb.append(0);
            }
            sb.append(hv);
            if (i == 3 | i == 5 | i == 7 | i == 9) {
                sb.append("-");
            }
        }
        return sb.toString();
    }

    private static byte char2Byte(char c) {
        byte b = 0;
        switch (c) {
            case '0':
                b = 0;
                break;
            case '1':
                b = 1;
                break;
            case '2':
                b = 2;
                break;
            case '3':
                b = 3;
                break;
            case '4':
                b = 4;
                break;
            case '5':
                b = 5;
                break;
            case '6':
                b = 6;
                break;
            case '7':
                b = 7;
                break;
            case '8':
                b = 8;
                break;
            case '9':
                b = 9;
                break;
            case 'a':
                b = 10;
                break;
            case 'b':
                b = 11;
                break;
            case 'c':
                b = 12;
                break;
            case 'd':
                b = 13;
                break;
            case 'e':
                b = 14;
                break;
            case 'f':
                b = 15;
                break;
            default:
                break;
        }
        return b;
    }

}

File: async/src/main/scala/com/fan/mysql/async/binlog/parse/PreviousGtidsParser.scala
package com.fan.mysql.async.binlog.parse

import java.math.BigInteger

import com.fan.mysql.async.binlog.event.impl.PreviousGtidsEvent
import com.fan.mysql.async.binlog.event.{BinlogEvent, EventHeader}
import com.fan.mysql.async.binlog.position.GtidSetPosition
import com.fan.mysql.async.binlog.{BinlogDumpContext, BinlogEventParser}
import com.fan.mysql.async.util.ByteUtil
import com.fan.mysql.async.util.ChannelWrapper._
import io.netty.buffer.ByteBuf

class PreviousGtidsParser extends BinlogEventParser {

  override def parse(buffer: ByteBuf, header: EventHeader, context: BinlogDumpContext): BinlogEvent = {
    val event = new PreviousGtidsEvent(header)
    val sb = new StringBuffer
    val sidNumberCount = buffer.readLong()
    var i = 0
    while ( {
      i < sidNumberCount
    }) {
      if (sb.length > 0) sb.append(",")
      val sourceId = buffer.readBytes(16).array()
      val uuidStr = parseServerId(sourceId)
      sb.append(uuidStr)
      val internalCount = buffer.readLong()
      var j = 0
      while ( {
        j < internalCount
      }) {
        val from = buffer.readUnsignedLong()
        val to = buffer.readUnsignedLong() - BigInteger.valueOf(1)
        sb.append(":").append(from).append("-").append(to)

        j += 1
      }

      i += 1
    }
    event.setGtidSet(sb.toString)
    event
  }

  def getBody(gtidSet: String): Array[Byte] = {
    val pos = new GtidSetPosition(gtidSet)
    // calculate body size
    var bodySize = 8
    val uuids = pos.getUUIDSets

    uuids.forEach { uuidSet =>
      bodySize += 16 + 8
      bodySize += 16 * uuidSet.getIntervals.size
    }
    // fill body
    val body = new Array[Byte](bodySize)
    var offset = 0
    ByteUtil.int8store(body, offset, uuids.size)
    offset += 8

    uuids.forEach { uuidSet =>
      val uuidByte = parseUuid(uuidSet.getUUID)
      System.arraycopy(uuidByte, 0, body, offset, 16)
      offset += 16
      ByteUtil.int8store(body, offset, uuidSet.getIntervals.size)
      offset += 8

      uuidSet.getIntervals.forEach { interval =>
        ByteUtil.int8store(body, offset, interval.getStart)
        offset += 8
        ByteUtil.int8store(body, offset, interval.getEnd + 1)
        offset += 8
      }
    }
    body
  }

  private def parseUuid(uuid: String) = {
    val source_id = new Array[Byte](16)
    var array_index = 0
    var str_index = 0

    import scala.util.control.Breaks._

    while (str_index < uuid.length) {
      breakable {
        val high_char = uuid.charAt({
          str_index += 1
          str_index - 1
        })
        if (high_char == '-') break() // continue
        val high = char2Byte(high_char)
        val low_char = uuid.charAt({
          str_index += 1
          str_index - 1
        })
        val low = char2Byte(low_char)
        source_id({
          array_index += 1
          array_index - 1
        }) = (high * 16 + low).toByte
      }
    }
    source_id
  }

  private def parseServerId(sourceId: Array[Byte]): String = {
    val sb = new StringBuilder
    if (sourceId == null || sourceId.length <= 0) return null
    for (i <- sourceId.indices) {
      val v = sourceId(i) & 0xff
      val hv = Integer.toHexString(v)
      if (hv.length < 2) sb.append(0)
      sb.append(hv)
      if (i == 3 | i == 5 | i == 7 | i == 9) sb.append("-")
    }
    sb.toString
  }

  private def char2Byte(c: Char) = {
    var b = 0
    c match {
      case '0' =>
        b = 0

      case '1' =>
        b = 1

      case '2' =>
        b = 2

      case '3' =>
        b = 3

      case '4' =>
        b = 4

      case '5' =>
        b = 5

      case '6' =>
        b = 6

      case '7' =>
        b = 7

      case '8' =>
        b = 8

      case '9' =>
        b = 9

      case 'a' =>
        b = 10

      case 'b' =>
        b = 11

      case 'c' =>
        b = 12

      case 'd' =>
        b = 13

      case 'e' =>
        b = 14

      case 'f' =>
        b = 15

      case _ =>

    }
    b
  }
}
